src/compiler/parser.ts

```diff
@@ -3160,6 +3160,11 @@ module ts {
                 return undefined;
             }
 
+            // Must be no line terminator before token `=>`.
+            if (scanner.hasPrecedingLineBreak()) {
+                return undefined;

```

Comment:
Correct.  I would not do this here.  I would detect and parse out the arrow function, but then report later that it isn't valid.  This will help ensure that all the higher layers of the stack properly understand what's going on.